libs/mock-builder/src/lib.rs

```diff
@@ -39,177 +39,72 @@
 //!
 //! Take a look to the [pallet tests](`tests/pallet.rs`) to have a user view of how to use this crate.
 
-/// Provide methods for register/execute calls
-pub mod storage {
-	use std::{any::Any, cell::RefCell, collections::HashMap};
+/// Provide functions for register/execute calls
+pub mod storage;
 
-	/// Identify a call in the call storage
-	pub type CallId = u64;
-
-	trait Callable {
-		fn as_any(&self) -> &dyn Any;
-	}
-
-	thread_local! {
-		static CALLS: RefCell<HashMap<CallId, Box<dyn Callable>>>
-			= RefCell::new(HashMap::default());
-	}
-
-	struct CallWrapper<Input, Output>(Box<dyn Fn(Input) -> Output>);
-
-	impl<Input: 'static, Output: 'static> Callable for CallWrapper<Input, Output> {
-		fn as_any(&self) -> &dyn Any {
-			self
-		}
-	}
-
-	/// Register a call into the call storage.
-	/// The registered call can be uniquely identified by the returned `CallId`.
-	pub fn register_call<F: Fn(Args) -> R + 'static, Args: 'static, R: 'static>(f: F) -> CallId {
-		CALLS.with(|state| {
-			let registry = &mut *state.borrow_mut();
-			let call_id = registry.len() as u64;
-			registry.insert(call_id, Box::new(CallWrapper(Box::new(f))));
-			call_id
-		})
-	}
-
-	/// Execute a call from the call storage identified by a `call_id`.
-	pub fn execute_call<Args: 'static, R: 'static>(call_id: CallId, args: Args) -> R {
-		CALLS.with(|state| {
-			let registry = &*state.borrow();
-			let call = registry.get(&call_id).unwrap();
-			call.as_any()
-				.downcast_ref::<CallWrapper<Args, R>>()
-				.expect("Bad mock implementation: expected other function type")
-				.0(args)
-		})
-	}
-}
+/// Provide functions for handle fuction locations
+pub mod location;
 
+use frame_support::{Blake2_128, StorageHasher, StorageMap};
+use location::FunctionLocation;
 pub use storage::CallId;
 
 /// Prefix that the register functions should have.
 pub const MOCK_FN_PREFIX: &str = "mock_";
 
-/// Gives the absolute string identification of a function.
-#[macro_export]
-macro_rules! function_locator {
-	() => {{
-		// Aux function to extract the path
-		fn f() {}
-
-		fn type_name_of<T>(_: T) -> &'static str {
-			std::any::type_name::<T>()
-		}
-		let name = type_name_of(f);
-		&name[..name.len() - "::f".len()]
-	}};
+/// Register a mock function into the mock function storage.
+/// This function should be called with a locator used as a function identification.
+pub fn register<Map, L, F, I, O>(locator: L, f: F)
+where
+	Map: StorageMap<<Blake2_128 as StorageHasher>::Output, CallId>,
```

Comment:
I don't see right now a use case where a caller would want to choose another `Hasher` in this case. 

But thanks for your thoughts! I keep that in mind in case it evolves in a way a different `Hasher` can be useful.